Critical adjustments, Manpower into Food changes and some bugfixes

Ninakoru opened the pull request:
Lowers crit chance in some outliers.
Energy/Projectile enhancers. crit chance is now for energy/projectile weapon types.
Fixed Nakalim Fleet Enhancer not restricted to only one of that type. For Cravers version I didn't need to update the restriction.

As agreed in Issue: #1584

Commit:
Critical hit chance adjustments

wakystuf commented:
This looks good to me. One question: will this path end up applying to modules other than missiles and slugs (e.g. railguns, BEB, squadrons)?

<Modifier TargetProperty="CriticalHitChance" Operation="Addition" Value="$(CriticalHitChanceProjectile)" Path="ClassShip/ClassSection/ClassModuleWeapon,!ClassModuleWeaponLaser,!ClassModuleWeaponBeam" Priority="1" />

Ninakoru commented:
> This looks good to me. One question: will this path end up applying to modules other than missiles and slugs (e.g. railguns, BEB, squadrons)?

Well, If it didnt before, sure will not do now. The old modifier was to all ClassModuleWeapon.
Supposedly if a special weapon does not use crit, either doesn't have the crit property or have a Force to 0.
I used that specific way to make sure if something had crit before, still gets it from projectile enhancer (for mods and anything)

Commit:
Dynamic manpower from food and fix of vanilla bug.

There it goes, I did a lot of testing in a modded vanilla version, but also a fair bit in ESG, so it should be stable a working with no issues.

The bug is now fixed, this means you don't get free refills, this happened when your Empire Manpower was full and it changed briefly into full conversion, did the refills for free, as the food was already computated in systems.

Now aside the food computation, the conversion you see on the tooltip will be stored internally and used on the Empire Manpower computations. What you see is what you get next turn, and every bit of manpower will be properly taken from your manpower sources (mainly food conversion and minor civs ^^).

On the bright side we have the dynamic ratio of food sent. If you need to cover 400 manpower and you can produce 2000 from food, only 20% will be converted from food, keeping the manpower from food expenditures more fluid, aside the 5% overcorrection, that will mean in practice much, much less wasted overall.

About the code:

The meat and potatoes is on ResourceDefinitions, SimulationDescriptors[ColonizedStarSystem] and SimulationDescriptors[Empire]. All the other changes are mostly due to the need or an alternative Manpower property, EmpireManpowerBalance, to avoid any kind of circular loops, also EmpireManpowerPercent on the three effects that use percents instead of additions.

In ResourceDefinitions, we added a new stock resource, for internal purposes only: store the food conversion used when you clicked end of turn. TrueNetEmpireManpower is the new NetEmpireManpower operating behind the scenes.

In SimulationDescriptors[ColonizedStarSystem] "ShouldTakeGrowthForManpower" gets replaced with GrowthForManpowerRatio, the first was the on-off switch from vanilla, the latter will storage the percentage used for the dynamic conversion, at the same moment food is transferred. The maximum capacity of conversion is calculated here, and the ratio is applied also here.

SimulationDescriptors[Empire]: We get a bunch of new properties for internal calculations most of it: MaximumEnrollmentRate gets the feedback form colonies using the Composition="Sum" property, the storage of the conversion quantity used is done here, also the calculation of the ratio of conversion needed to cover the missing empire manpower.

captaincobbs commented:
Includes #1584 and #1585

wakystuf commented:
OK, that sounds good on the crit changes. 

I will also look at the commit for the MP changes, which may take me a bit longer to carefully review, as they are more complex.

Commit:
Fix to action declare war to minor, and adjustment on MP calculation.

Fixed the declare war to minor launching an exception.

Restricted the food accumulator and the later calculations into True Net Manpower to Major factions, as the Resource is only available for majors (not a big deal, but anyways).

Ninakoru commented:
Added the fix for #1588, also a little update on #1585 (not a big deal)

BTW I cannot add the tags on the issues :(

Commit:
Fix to bug of Serious difficulty and minor faction.

Adds the restriction so minor factions don't get 300 empire manpower generation.

Fixes: #1589

Also added a missing modifier so manpower rate also is properly calculated with frozen systems due to Riftborn ultimate singularity (quite specific).

Fixes: #1585

wakystuf commented:
I DM'd you a few questions on the MP/food mechanics. Everything else looks great.

wakystuf commented:
OK, I think this is more or less all set, though I'd like to do a bit of testing if I can find the time. Open items:
1) Carefully watch the (vanilla) EmpireManpowerUpkeep variable to make sure it is working properly; I think it may not be, but it's opaque (sealed) and therefore hard to tell for sure.
2) Account for the various random sources of flat system MP in the proportional enrollment calculation
Bonus: Explicate the GUI on the main galaxy screen (not the military screen, just the one if you hover over MP)

Commit:
UI Feedback update for manpower tooltips

Added "(Manpower) from Fleets/Systems" and "(Manpower) from Lifeforce to Manpower" (with icons) for those two components of the manpower breakdown in the tooltip of Empire Manpower.

Minor QoL update for #1585

wakystuf commented:
All looks good, so I merged. I'm going to try to test a little bit before I push to Steam
